Enable Swift C++ interop in WebGPU.framework.

nmahendru opened the pull request:
#### fb45993e8c151ffe248aa67fb775667021afb1c5
<pre>
Enable Swift C++ interop in WebGPU.framework.
<a href="https://rdar.apple.com/135914808">rdar://135914808</a>

Reviewed by Geoffrey Garen and Mike Wyrzykowski.

This change will serve as an exemplar for using Swift Code in WebGPU.framework.
The swift generated header has been marked no-install to intentionally keep it hidden
and not expose from the WebGPU.framework.

The actual change of moving the RemoteBuffer::copy operation to swift has no relevance
per se. It&apos;s a randomly chosen operation to use a swift API.

As a result of this change, the WebGPU module(WebGPU.modulemap) is consumed by the swift compiler for consuming
WebGPU types in Swift if needed.

Swift compiler is aliasing both UChar and uint16_t to unsigned short and thus it&apos;s not able to compile certain utilities
that are ifdefed out.

The changes are disabled and manual testing has been performed at desk by:
1. make release BUILD_WEBKIT_OPTIONS=&quot;--webGpuSwift&quot;
2. Running ./Tools/Scripts/run-webkit-tests --release --no-build --force http/tests/webgpu/webgpu/api/operation/buffers http/tests/webgpu/webgpu/api/validation/buffer

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::copy):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::copy):
(wgpuBufferCopy):
* Source/WebGPU/WebGPU/Buffer.swift: Added.
(copySpan(_:source:)):
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebGPU/WebGPU/WebGPU.modulemap:
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebGPU/WebGPU/WebGPUSwift.h: Added.
* Source/WebGPU/WebGPU/WebGPUSwiftInternal.h: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copy):
* Tools/Scripts/webkitperl/FeatureList.pm:

Canonical link: <a href="https://commits.webkit.org/284166@main">https://commits.webkit.org/284166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/231085e5c7fd12897b15c186be55584863aea797

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72634 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70682 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54696 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13111 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43830 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59201 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40497 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16624 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18067 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61683 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62450 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74328 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67813 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12536 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16226 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62172 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12576 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62198 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10125 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3741 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89592 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10452 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43758 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15867 "Found 8 new JSC stress test failures: wasm.yaml/wasm/function-tests/many-arguments-to-function.js.wasm-collect-continuously, wasm.yaml/wasm/fuzz/memory.js.wasm-bbq, wasm.yaml/wasm/fuzz/memory.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-collect-continuously, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/stress/tail-call.js.default-wasm, wasm.yaml/wasm/stress/tail-call.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44832 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44574 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -913,7 +913,7 @@ inline void copyElements(ElementType* __restrict destination, const ElementType*
 }
 
-inline void copyElements(uint16_t* __restrict destination, const uint8_t* __restrict source, size_t length)
+inline void copyElements16To8(uint16_t* __restrict destination, const uint8_t* __restrict source, size_t length)
```
When we have correct code, but we need to suppress it in order to work around a limitation in a particular compiler, the best approach is to #ifdef or #ifndef for that compiler.

You can see this in cases like #ifdef __cplusplus, #ifdef __OBJC__, #if !COMPILER(MSVC), and so on.

#ifdef for the errant or un-supporting compiler is more sustainable in a project that supports many compilers, and it's more self-documenting.

geoffreygaren commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -294,7 +294,13 @@ template<> class StringTypeAdapter<Unicode::CheckedUTF8, void> {
+    // FIXME: Only added for Swift C++ interop and UChar is always 16 bits so the bitwise_cast is fine.
+    // The real issue is that Swift considers char16_t as an alias for unsigned short, though char16_t is a type as of C++20. See rdar://136156228.
+    void writeTo(UChar* destination) const { Unicode::convert(m_characters.characters, std::span { bitwise_cast<char16_t *>(destination), m_characters.lengthUTF16 }); }
+#endif
```
Do we need to call this from Swift? If not, probably better not to #define a Swift alternative.

geoffreygaren commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -561,10 +561,14 @@ inline String operator"" _str(const char* characters, size_t)
 }
 
+#ifndef __swift__
+// FIXME: Interop does not import operators. Moreover, this fails to compile because of rdar://136156228
```
`#ifndef __swift__ // FIXME: rdar://136156228` can suffice.

geoffreygaren commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
```diff
@@ -96,9 +96,11 @@ std::span<uint8_t> BufferImpl::getBufferContents()
+    if (backing())
+        return wgpuBufferCopy(backing(), data, offset);
+    return false;
 }
```
Why did we need to add an implementation here? (Does the implementation work? What config tests it?)

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.swift:
```diff
@@ -0,0 +1,34 @@
+
+public class BufferUtils {
+    public static func copySpan(destination: SpanUInt8, source: SpanConstUInt8) -> Bool {
+        if source.size_bytes() > destination.size_bytes() {
```
We want precondition() here, rather than a return value. In general, we don't want to change the structure or semantics of existing code when we add Swift code. Also, the caller has no meaningful way to recover from failure.

nmahendru replied:
good suggestion. Will use that.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.swift:
```diff
@@ -0,0 +1,34 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+public class BufferUtils {
```
I don't think a generic namespace helps here. We can make this an extension to SpanUInt8, or just a free function like memcpySpan.

geoffreygaren commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
```diff
@@ -98,21 +98,7 @@ void RemoteBuffer::copy(std::optional<WebCore::SharedMemoryHandle>&& dataHandle,
-    }
-
-    memcpySpan(buffer.subspan(offset), data);
-    completionHandler(true);
```
Here again, we want to avoid changing the structure / design as much as possible when we refactor. 

Let's keep the existing structure, and just sub in the other copy function if we're building with the Swift prototype enabled.

nmahendru replied:
With this PR we are only enabling Swift interop at the WebGPU.framework level. Not at the WebKit Level. So I can't stub out the Swift API here.
The intent here is to move the copy operation down to that level. So, while doing that I am also moving bounds checking down so that we check at the same level we call memcopy, it's just more coherent.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -479,6 +482,25 @@ static size_t computeRangeSize(uint64_t size, size_t offset)
+    memcpySpan(std::span<uint8_t> { buffer + offset, currentSize() - offset }, data);
+    return true;
+#else
+    return WebGPU::BufferUtils::copySpan(std::span<uint8_t> { buffer + offset, currentSize() - offset }, data);
```
We want to structure our Swift prototype in WebGPU a little differently. Instead of unconditionally turning on the prototype when it's available, we want a separate #ifdef that controls it. This is a great patch to introduce the separate #ifdef.

The reasons we want a separate #ifdef are:
* We want to be able to A/B test prototype vs shipping for performance and correctness
* The prototype will soon use Swift features that aren't available in any shipping Swift compiler

geoffreygaren reviewed (commented):
This is a good start. Some suggestions to change the structure a bit.

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
```diff
@@ -98,6 +98,7 @@ void RemoteBufferProxy::copy(std::span<const uint8_t> span, size_t offset)
         RELEASE_ASSERT(sharedMemory.get() || !handleHasValue);
     });
     UNUSED_VARIABLE(sendResult);
+    return;
```
Change is not needed afaik, but no objection to it either

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1862,7 +1865,8 @@ WGPU_EXPORT void wgpuXRViewRelease(WGPUXRView view) WGPU_FUNCTION_ATTRIBUTE;
 
 #endif  // !defined(WGPU_SKIP_DECLARATIONS)
-
+using SpanConstUInt8 = std::span<const uint8_t>;
```
This probably needs to be under a `ifdef _cplusplus` otherwise the SDK verifier will complain, unless the SDK verifier was complaining due to `[extern_c]` in the module map.

mwyrzykowski reviewed (approved):
I think we need to move the `span` usage under `ifdef __cpluscplus` in WebGPU.h but otherwise looks great

geoffreygaren commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1074,6 +1074,9 @@
 
 #if PLATFORM(COCOA)
 #define HAVE_WEBGPU_IMPLEMENTATION 1
+#if !defined(HAVE_WEBGPU_SWIFT)
```
Let's use ENABLE here. ENABLE will sometimes be false even on platforms that HAVE the capability.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -479,6 +482,21 @@ static size_t computeRangeSize(uint64_t size, size_t offset)
+    WTFLogAlways("NITIN %s:%d", __FILE__, __LINE__);
+    WebGPU::copySpan(destination, data);
+}
+#endif
```
It's fine for this patch to be minimal -- but the obvious next step is to move this code into the Swift WebGPU module, so that you get all the null checking and overflow checking automatically.

geoffreygaren reviewed (approved):
r=me

nmahendru committed:
Enable Swift C++ interop in WebGPU.framework.
rdar://135914808

Reviewed by Geoffrey Garen and Mike Wyrzykowski.

This change will serve as an exemplar for using Swift Code in WebGPU.framework.
The swift generated header has been marked no-install to intentionally keep it hidden
and not expose from the WebGPU.framework.

The actual change of moving the RemoteBuffer::copy operation to swift has no relevance
per se. It's a randomly chosen operation to use a swift API.

As a result of this change, the WebGPU module(WebGPU.modulemap) is consumed by the swift compiler for consuming
WebGPU types in Swift if needed.

Swift compiler is aliasing both UChar and uint16_t to unsigned short and thus it's not able to compile certain utilities
that are ifdefed out.

The changes are disabled and manual testing has been performed at desk by:
1. make release BUILD_WEBKIT_OPTIONS="--webGpuSwift"
2. Running ./Tools/Scripts/run-webkit-tests --release --no-build --force http/tests/webgpu/webgpu/api/operation/buffers http/tests/webgpu/webgpu/api/validation/buffer

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::copy):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::copy):
(wgpuBufferCopy):
* Source/WebGPU/WebGPU/Buffer.swift: Added.
(copySpan(_:source:)):
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebGPU/WebGPU/WebGPU.modulemap:
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebGPU/WebGPU/WebGPUSwift.h: Added.
* Source/WebGPU/WebGPU/WebGPUSwiftInternal.h: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copy):
* Tools/Scripts/webkitperl/FeatureList.pm:

Canonical link: https://commits.webkit.org/284166@main
